### PR TITLE
build-providers: add support for --shell-after 

### PR DESCRIPTION
**Pull request by sergiusens**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/cli/lifecycle.py [hunk 1]
```diff
@@ -65,6 +66,8 @@ def _execute(
                     instance.pack_project(output=output)
                 else:
                     instance.execute_step(step)
+                if shell_after:
+                    instance.shell()
```
Comment (chipaca) [inline snapcraft/cli/lifecycle.py]: This is a little surprising. What happens if the shell exits with error status, at  this point?
I'd expect the shell_after check to be done after we'd cleared the `try`/`except` block -- if necessary with its own exception handler.

Comment (sergiusens) [reply snapcraft/cli/lifecycle.py]: it will raise a proper exception that is captured correctly in our exit handler

Comment (chipaca) [reply snapcraft/cli/lifecycle.py]: is that what you want? the step itself didn't fail, just the shell after. Are the same messages still applicable? Do you want shell and debug to interact in this way?

Comment (sergiusens) [reply snapcraft/cli/lifecycle.py]: I see your point now. Will fix too

**Pull request merged**